File: demo/curve1.c
/*
 * Library:  lmfit (Levenberg-Marquardt least squares fitting)
 *
 * File:     demo/curve1.c
 *
 * Contents: Example for one-dimensional curve fitting,
 *           using the simplified interface from lmcurve.h.
 *
 * Author:   Joachim Wuttke <j.wuttke@fz-juelich.de> 2004-2013
 * 
 * Licence:  see ../COPYING (FreeBSD)
 * 
 * Homepage: apps.jcns.fz-juelich.de/lmfit
 */
 
#include "lmcurve.h"
#include <stdio.h>

/* model function: a parabola */

double f( double t, const double *p )
{
    return p[0] + p[1]*t + p[2]*t*t;
}

int main()
{
    int n = 3; /* number of parameters in model function f */
    double par[3] = { 100, 0, -10 }; /* really bad starting value */
    
    /* data points: a slightly distorted standard parabola */
    int m = 9;
    int i;
    double t[9] = { -4., -3., -2., -1.,  0., 1.,  2.,  3.,  4. };
    double y[9] = { 16.6, 9.9, 4.4, 1.1, 0., 1.1, 4.2, 9.3, 16.4 };

    lm_status_struct status;

    printf( "Fitting ...\n" );
    lmcurve( n, par, m, t, y, f, &lm_control_double, NULL, &status );
        
    printf( "Results:\n" );
    printf( "status after %d function evaluations:\n  %s\n",
            status.nfev, lm_infmsg[status.info] );

    printf("obtained parameters:\n");
    for ( i = 0; i < n; ++i)
        printf("  par[%i] = %12g\n", i, par[i]);
    printf("obtained norm:\n  %12g\n", status.fnorm );
    
    printf("fitting data as follows:\n");
    for ( i = 0; i < m; ++i)
        printf( "  t[%2d]=%4g y=%6g fit=%10g residue=%12g\n",
                i, t[i], y[i], f(t[i],par), y[i] - f(t[i],par) );

    return 0;
}

File: pythondemo/curve1.py
import swig.pylmfit as pylm

def f(t, p):
    return p[0] + p[1]*t + p[2]*t**2

def curve1():
    par0 = pylm.VectorDouble([100, 0, -10])
    t = pylm.VectorDouble([-4, -3, -2, -1, 0, 1, 2, 3, 4])
    y = pylm.VectorDouble([16.6, 9.9, 4.4, 1.1, 0, 1.1, 4.2, 9.3, 16.4])
    control = pylm.lm_control_double
    pylm.set_parlen(3)
    pylm.set_callback(f)

    print("Fitting...")
    result = pylm.fit_curve(par0, t, y, pylm.function_g, control)

    print("Results:")
    print(f"status after {result.status.nfev} function evaluations:")
    pylm.print_infmsg(result.status.outcome)

    print("Obtained parameters:")
    for j in range(pylm.parlen()):
        print(f"par[{j}] = {result.par[j]}")
    print("Obtained norm:")
    print(result.status.fnorm)

    print("fitting data as follows:")
    for i in range(t.size()):
        fi = f(t[i], result.par)
        print(f"t[{i}] = {t[i]}  y = {y[i]}  fit = {fi}  residue = {y[i]-fi}")

    if (result.status.outcome > 3):
        print("FAILURE")
        return 1
    print("SUCCESS")
    return 0

if __name__ == "__main__":
    print(curve1())
